ratelimits/gcra.go

```diff
@@ -0,0 +1,121 @@
+package ratelimits
+
+import (
+	"math"
+	"time"
+
+	"github.com/jmhodges/clock"
+)
+
+// divThenRound divides two int64s and rounds the result to the nearest integer.
+// This is used to calculate request intervals and costs in nanoseconds.
+func divThenRound(x, y int64) int64 {
+	return int64(math.Round(float64(x) / float64(y)))
+}
+
+// maybeSpend uses the GCRA algorithm to decide whether to allow a request. It
+// returns a Decision struct with the result of the decision and the updated
+// TAT. The cost must be 0 or greater and <= the burst capacity of the limit.
+func maybeSpend(clk clock.Clock, rl limit, tat time.Time, cost int64) *Decision {
+	if cost < 0 || cost > rl.Burst {
+		panic("invalid cost for maybeSpend")
+	}
+	nowUnix := clk.Now().UnixNano()
+	tatUnix := tat.UnixNano()
+
+	// If the TAT is in the future, use it as the starting point for the
+	// calculation. Otherwise, use the current time. This is to prevent the
+	// bucket from being filled with capacity from the past.
+	if nowUnix > tatUnix {
+		tatUnix = nowUnix
+	}
+
+	// Compute the cost increment.
+	emissionInterval := divThenRound(rl.Period.Nanoseconds(), rl.Count)
+	costIncrement := emissionInterval * cost
+
+	// Deduct the cost to find the new TAT and residual capacity.
+	newTAT := tatUnix + costIncrement
+	burstOffset := emissionInterval * rl.Burst
+	difference := nowUnix - (newTAT - burstOffset)
+	residual := divThenRound(difference, emissionInterval)
+
+	if costIncrement <= 0 && residual == 0 {
```

Comment:
> It seems like this case should be impossible, based on the fact that cost, count, and period must all be positive.

The only guarantee made is that cost is equal or greater then zero. A zero cost maybeSpend() via Limiter.Check() is used in our tests to check bucket state after in many of our unit tests. This will continue to be an important feature used in tests outside of this package and any subscriber facing rate limits API.

The condition `costIncrement <= 0 && residual == 0` is true when a zero cost Limiter.Check() is performed on an empty bucket. There is coverage for this case in the unit tests. I'll add a comment to indicate this.

Reply:
I've actually managed to convince myself that a 0 cost request is technically allowed and so allowed should be true. When I originally wrote this Spend called Check and used `Decision.Allowed` to indicate that the TAT should be updated, however this is no longer the case (for good reason).